[compiled autograd] torch.compile API

xmfan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #125880

- enter existing compiled autograd ctx manager before entering torch.compile frames

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

xmfan committed:
[wip][compiled autograd] torch.compile API

[ghstack-poisoned]

xmfan committed:
Update on "[wip][compiled autograd] torch.compile API"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

xmfan committed:
Update on "[wip][compiled autograd] torch.compile API"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

xmfan committed:
Update on "[wip][compiled autograd] torch.compile API"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

xmfan committed:
Update on "[wip][compiled autograd] torch.compile API"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

xmfan committed:
Update on "[wip][compiled autograd] torch.compile API"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

xmfan committed:
Update on "[wip][compiled autograd] torch.compile API"


torch.compile API:
- torch._dynamo.config flag for compiled autograd, controls whether or not we mark autograd nodes during forward pass
- forward compiler options to compiled autograd using a rewrap fn
- forward via TracingContext and AOTAutograd CompiledFunction node
- autograd engine traverses autograd graph, and routes to compiled autograd if it finds a marked node

keep previous enable/disable context manager api for tests + fine-grained use cases:
- override c++ autograd engine to route to compiled autograd or eager
- does not change global config flag

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

xmfan committed:
Update on "[compiled autograd] torch.compile API"


torch.compile API:
- torch._dynamo.config flag for compiled autograd, controls whether or not we mark autograd nodes during forward pass
- forward compiler options to compiled autograd using a rewrap fn
- forward via TracingContext and AOTAutograd CompiledFunction node
- autograd engine traverses autograd graph, and routes to compiled autograd if it finds a marked node

keep previous enable/disable context manager api for tests + fine-grained use cases:
- override c++ autograd engine to route to compiled autograd or eager
- does not change global config flag

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

jansel commented on torch/__init__.py:
```diff
@@ -1898,14 +1901,24 @@ def fn(model: Callable):
                            disable=disable)
         return fn
 
+    def rewrap(model):
+        return functools.partial(
+            compile,
+            fullgraph=fullgraph,
+            dynamic=dynamic,
+            backend=backend,
+            mode=mode,
+            options=options,
+            disable=disable)(model)
```
I'm a bit confused by the usage of partial here, perhaps you want:
```suggestion
    rewrap = functools.partial(
            compile,
            fullgraph=fullgraph,
            dynamic=dynamic,
            backend=backend,
            mode=mode,
            options=options,
            disable=disable)
```

jansel commented on torch/_dynamo/compiled_autograd.py:
```diff
@@ -269,35 +286,59 @@ def set_node_origin(self, node_name, node_index):
-    global compiled_autograd_enabled, override_compiler_fn
-    assert not compiled_autograd_enabled
-    compiled_autograd_enabled = True
+    global override_compiler_fn
```
Why is this called "override" are we overriding some other setting?  Or is this just the main way to set it?

xmfan replied:
suppose we have 
```python
torch._dynamo.config.compiled_autograd = True
compiled_fn = torch.compile(fn, backend="inductor")  # _initialize()
loss = compiled_fn()
with enable(torch.compile(backend="aot_eager")):
  loss.backward()
```

backend aot_eager overrides the inductor compile wrapper set on the autograd nodes

jansel commented on torch/_dynamo/config.py:
```diff
@@ -437,6 +437,11 @@ def default_debug_dir_root():
+# This flag is used during forward compilation to prepare compile modified autograd nodes
+# for compiled autograd. To disable compiled autograd after the forward compilation, use the
+# torch._dynamo.compiled_autograd.disable() context manager instead
+compiled_autograd_enabled = False
```
```suggestion
compiled_autograd = False
```

jansel commented on torch/_dynamo/eval_frame.py:
```diff
@@ -668,6 +668,8 @@ def toy_example(a, b):
     if disable or os.environ.get("TORCHDYNAMO_DISABLE", "") == "1":
         return _NullDecorator()
+    elif config.compiled_autograd_enabled:
+        torch._dynamo.compiled_autograd._initialize()
```
Wont installing compiled autograd globally like this has bad global side effects?

If I have two models, one compiled and one eager -- then this will enable compiled autograd for the unrelated model.

xmfan replied:
I don't think it should? I changed the autograd engine routing logic in engine.cpp:1270. it now needs
- a compiled being set (by _initialize)
- either
  - an autograd graph containing a CompiledFunctionBackward marked by compiled autograd
  - .backward called within the ctx manager override

the eager model's output on which we call .backward shouldn't have an autograd graph containing CompiledFunctionBackward, so autograd engine would route to eager.

jansel reviewed (changes requested):
Test failures

albanD reviewed (commented):
I'm a bit surprised that torch.compile now has effect at a distance where code unrelated to the fw/bw it captures will also run under compile?
I was under the expectation that only .backward() calls within torch.compile would trigger compiled autograd?
Can you share some details on why this has changed?

xmfan commented:
@albanD Could you elaborate more? If some tensor's autograd graph contains a node created by torch.compile (e.g. CompiledFunctionBackward), it should run compiled autograd.

The context manager API that would let users to do eager fw + CA bw is there mostly for tests

albanD commented:
> If some tensor's autograd graph contains a node created by torch.compile (e.g. CompiledFunctionBackward), it should run compiled autograd.

My question is why? This custom Function already properly contains the compiled backward for the op that was marked for compilation during the forward.
But enabling CA overall here would trigger compilation for the backward of ops that were not in the compiled region in the forward right?

xmfan commented:
Yes, CA tries to compile the entire backward, even the nodes originating from fw graph break regions. Without this, we would have equal sized graphs as AOTAutograd, and can't capture DDP/FSDP hooks

albanD commented:
> If I understand the concern properly, CA should only handle gradients for nodes originating between the compiled start and compiled end regions in this example:

My mental model of what happens here is that we only ever compile code that is explicitly wrapped via torch.compile.
So doing torch.compile of a function in the forward will compile that function and the corresponding backward.
If the user wants the whole backward to be compiled, they should wrap the call to .backward() within torch.compile to state that they want all of it to be compiled.

I find the current logic that will trigger compiled backward beyond what is torch.compile-d explicitly surprising. And I feel like a lot of the complexity of this PR comes from this: We have to leave breadcrumbs in the autograd.Function generated by AOTAutograd so that we can then traverse the whole autograd graph and find such a Node to then trigger a full compile.
This sounds very complex and brittle compared to `torch.compile(lambda: loss.backward())()` that I would expect the user to do if they want their whole backward to be compiled in a single block.

xmfan commented:
That makes a lot of sense to me, I'll explore the implementation

xmfan committed:
Update

[ghstack-poisoned]

xmfan committed:
Update

[ghstack-poisoned]

xmfan committed:
Update on "[compiled autograd] torch.compile API"


- enter existing compiled autograd ctx manager before entering torch.compile frames

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

xmfan commented:
@jansel could you take another look

xmfan commented:
@pytorchbot merge

atalman commented:
@pytorchmergebot revert -c nosignal -m "sorry your PR broke lint, need to revert"

xmfan committed:
Update on "[compiled autograd] torch.compile API"


- enter existing compiled autograd ctx manager before entering torch.compile frames

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

xmfan committed:
Update on "[compiled autograd] torch.compile API"


- enter existing compiled autograd ctx manager before entering torch.compile frames

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

xmfan commented:
@pytorchbot merge
